File: stack/second_next_greater/second_next_greater.js
var find_second_next_greater = function (nums) { // LeetCode Q.2454.
    let second_next_greater = Array(nums.length).fill(-1);
    let stack_1 = [], stack_2 = [], transporter = []; // Format: [idx, num].

    for (const [idx, num] of nums.entries()) {
        while (stack_2.length > 0 && num > stack_2.at(-1)[1]) {
            let past_idx = stack_2.pop()[0];
            second_next_greater[past_idx] = num;
        }

        while (stack_1.length > 0 && num > stack_1.at(-1)[1])
            transporter.push(stack_1.pop()); // Stack 1 to transporter.
        stack_1.push([idx, num]); // Now current num can enter stack 1.

        while (transporter.length > 0)
            stack_2.push(transporter.pop()); // Transporter to stack 2.
    }

    return second_next_greater;
};

File: docs/stack/2454_hard/second_next_greater.cpp
#include <vector>
#include <stack>
using namespace std;

vector<int> find_second_next_greater(vector<int> &nums) // LeetCode Q.2454.
{
    vector<int> secondNextGreater(nums.size(), -1);

    stack<pair<int, int>> stackOne, stackTwo; // Format: {num, idx}.
    vector<pair<int, int>> transporter;       // Format: {num, idx}.

    for (int idx = 0; idx < nums.size(); idx++)
    {
        while (!stackTwo.empty() && stackTwo.top().first < nums[idx])
        {
            int past_idx = stackTwo.top().second;
            secondNextGreater[past_idx] = nums[idx];
            stackTwo.pop();
        }

        while (!stackOne.empty() && stackOne.top().first < nums[idx])
        {
            transporter.push_back(stackOne.top()); // Keep decreasing monotonicity.
            stackOne.pop();
        }

        while (!transporter.empty())
        {
            stackTwo.push(transporter.back());
            transporter.pop_back();
        }

        stackOne.push({nums[idx], idx});
    }
    return secondNextGreater;
}